Merge editAll into integration

xaterz opened the pull request:
Implemented the '**Edit All Mountings and Quotas**' page for modifying multiple mountings and quotas at the same time. 

Features:
* To select module for editing, enter module code/name in autocomplete search box
* Module selected will then appear in the table
* When mounting/quota is modified, the affected cells will be highlighted red in real time
* When a module is removed from the table, the changes will be discarded
* Click save button to save all changes

Help me check that all features are working correctly.

tgqiang commented on components/handlers/tentative_module_mountings.py:
```diff
@@ -19,7 +19,8 @@ class Tentative(object):
         '''
             Full_mounting_plan is a list of 'subplans'
-            Each subplan is a list of 4 attributes (code, name, sem 1 mounting, sem 2 mounting)
+            Each subplan is a list of 6 attributes 
```
Pylint: Trailing whitespace

tgqiang commented on components/model.py:
```diff
@@ -118,6 +118,10 @@ def delete_fixed_mounting(code, ay_sem):
     return database.delete_fixed_mounting(code, ay_sem)
```
All functions missing doc-strings. Will not insist that doc-strings be added for now.

tgqiang commented on templates/index.html:
```diff
@@ -44,8 +44,8 @@ <h2 class="panel-title">
-                            <li>Change Preclusions/Prerequisites</li>
-                            <li>Edit Module</li-->
+                            <li>Change Preclusions/Prerequisites</li-->
+                            <li><a id="home-page" href="/editAll">Edit All Mountings and Quotas</a></li>
```
This addition here does not seem to be accounted for in `test_app.py`.

**Please add a test case for navigating to `/editAll`.**

tgqiang commented on test/test_edit_all_mountings_and_quotas.py:
```diff
@@ -0,0 +1,80 @@
+            Tears down 'app.py' fixture and logs out
+        '''
+        session.tear_down(self.test_app)
+
```
You might want to also include test cases where certain buttons are clicked. (e.g. the Save button, and Reset Table button, etc).

See if there are any ways to test those 2 buttons with [paste](https://chromium.googlesource.com/external/trace-viewer/+/bf55211014397cf0ebcd9e7090de1c4f84fc3ac0/third_party/Paste/paste/fixture.py). If this cannot be done, then this test case is alright as it is.

tgqiang commented on templates/editAll.html:
```diff
@@ -0,0 +1,252 @@
+        document.getElementById(id).disabled = !state;
+    }
+
+    //Add red highlight to mounting and quota if mounting is changed
```
Steps taken:
1. Enter `/editall`
2. Select a module to edit mounting/quota for (I used CS1020)
3. Change Sem 1's quota to -1.
4. Press the Save button
5. Alert prompts 'Are you sure you want to save your changes?'. That invalid input's cell is shaded red.
6. Press OK.

Results/Comment:
1. Using red color for highlighting changes **was very misleading**. I thought there was something wrong with my inputs or something whenever I applied a change to an input (especially when I checked/unchecked the mounting checkbox). Red color *does* bring attention to something, but more often than not it signifies danger or something going wrong. Consider using another better color instead.

2. Modifying quota for a module did not cause the color change to be applied in real time. The color change was only triggered only after I pressed the Save button, which did not match the statement you made that affected cells are highlight in red in real time.

3. When I entered the invalid quota and I pressed 'OK' after the confirmation dialog, the page does nothing without explicit feedback (on why the page is not performing any action). To us it *is* obvious that the page should do nothing after the attempt to submit data with invalid fields, but this SHOULD be an explicit visual feedback to the user nevertheless.

4. Users should probably not be redirected to a different page after a successful attempt to do mass mounting/quota editing. I think it is better to let the users choose to navigate elsewhere themselves if they personally think that they are done with the edit-all-mountings-and-quotas page.

5. Also, please **do not** add JS scripts directly into the HTML page. The reason I request this is because it makes maintaining all the JavaScripts easier with (a) central JS file(s). Should you require targeting specific elements, you can attach classes/IDs to them and use [DOM methods](https://www.w3schools.com/jsref/dom_obj_all.asp). However, if you have strong justification for doing so, then it is totally fine.

tgqiang commented on templates/editAll.html:
```diff
@@ -0,0 +1,252 @@
+            }
+        })
+    });
+
```
Your implementation seems to be showing/hiding module rows, then iterating through the full list of modules and making the edit if the edited flag is set. This can be expensive in the long run.

You could first store a list of modules and their associated data when this page loads, then only pass back those data which were modified back to the backend handler.

tgqiang commented on test/test_edit_all_mountings_and_quotas_backend.py:
```diff
@@ -0,0 +1,340 @@
+        test_data = {}
+        test_data[test_module_code+'_isEdited'] = "True"
+        test_data[test_module_code+'_Sem1Mounting'] = test_mounting_1
+        #test_data[test_module_code+'_Sem1Quota'] = test_quota_1   #modified to empty
```
[TO-FIX]

```
    # Modify Sem 1 and 2 quotas
    test_data = {}
    ...
    ....
    #test_data[test_module_code+'_Sem1Quota'] = test_quota_1   #modified to empty
    ....
```
Why is this one commented out?

tgqiang replied:
There are other instances of commented-out lines in this suite of test cases. Seeing they are commented out, why are they made so? Are those needed or not needed? If needed, please make it clear on why these commented-out lines should be there. If not, please remove them.

If not commenting these lines *do* actually break something, please fix the bugs.

xaterz commented:
Most of the issues raised have been fixed. Please review my latest commits.

Issues that I do not think need to be addressed:
> All functions missing doc-strings. Will not insist that doc-strings be added for now.

model.py is now a facade object. The actual functions and their docstrings can be found in either database.py or helper.py.
I don't recommend having the docstrings be duplicated in model.py, because it may lead to inconsistencies in the case where someone edits a docstring in one file, but forgets to edit the other.

> Your implementation seems to be showing/hiding module rows, then iterating through the full list of modules and making the edit if the edited flag is set. This can be expensive in the long run.
>
> You could first store a list of modules and their associated data when this page loads, then only pass back those data which were modified back to the backend handler.

If I want to only pass back data that were modified, I would either have to 1) iterate through all data again to find those that are modified, which is not desirable, or 2) dynamically add/remove elements from the html form, which may be time-consuming to implement

Also, I really don't think that a simple O(n) check of a boolean flag is a very expensive process, even if we add all the modules of NUS into our system (about 5000 modules).

a0129998 commented:
suggestion:

![image](https://cloud.githubusercontent.com/assets/12403982/24597810/a0aeaa58-187a-11e7-8e81-9894cbf9e111.png)

1. add a description to emphasize to the user that they need to select from a drop down.
It may be confusing to them otherwise.

2. coordinate with @nlzz22 to set an upper bound for the quota.

3. warn the users with `refresh frequently,someone might be editing this module from another page` as the latest changes will overwrite all previous changes.

xaterz commented:
1. Fixed
2. Fixed
3. As long as the user does not refresh the page, any new changes made in other pages will **not** cause the data on this page to be updated, since all the data are pre-fetched when the page is loaded. I'm guessing you pressed the 'reset table' button (which actually refreshes the page), which was why you saw the data being updated.
